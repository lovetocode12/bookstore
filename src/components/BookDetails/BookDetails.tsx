import React, { useEffect, useContext } from 'react'
import { useParams } from 'react-router-dom'
import { useSelector } from 'react-redux';
import {
    ConatinerDiv, PhotoDiv, Image, ContentDiv,
    H2Tag, PTag, StyledButton, ButtonsContainer
} from './BookDetails_Style'
import useBuyBook from '../../common/useBuyBook';
import { Order } from '../../store/orders/type';
import { RootState } from '../../store/type';
import { TitleContext } from '../../common/TitleContext';

type BookParams = {
    book_id: string
}

type Props = {};

const BookDetails: React.FC<Props> = () => {

    const params: BookParams = useParams();

    const [{ author, url, title, description, price, pageCount, id }] = useSelector((state: RootState) => state.bookState.books.filter((book) => book.id === params.book_id))

    const [BuyNow, AddToCart] = useBuyBook();

    const context: any = useContext(TitleContext);
    const dispatchTitle = context.dispatchTitle;
    useEffect(() => {
        if (dispatchTitle) {
            dispatchTitle({ type: 'SET_TITLE', payload: title })
        }
    }, [dispatchTitle, title])

    function BuyNowClick() {
        let order: Order = {
            orderStatus: 'InProcess',
            orderDate: new Date(),
            book: {
                id, description,
                url, title, price, author, pageCount
            }

        }
        BuyNow(order);
    }

    function AddToCartClick() {
        let order: Order = {
            orderStatus: 'InProcess',
            orderDate: new Date(),
            book: {
                id, description,
                url, title, price, author, pageCount
            }

        }
        AddToCart(order);
    }


    return (
        <ConatinerDiv>
            <PhotoDiv >
                <Image src={url} alt={title} ></Image>
            </PhotoDiv>
            <ContentDiv>
                <div>
                    <H2Tag>{title}</H2Tag>
                </div>
                <div>
                    {price}
                </div>
                <div>
                    {author}
                </div>
                <div>
                    {pageCount}
                </div>
                <ButtonsContainer>
                    <StyledButton onClick={AddToCartClick}>Add To Cart</StyledButton>
                    <StyledButton onClick={BuyNowClick}>Buy Now</StyledButton>
                </ButtonsContainer>
                <PTag> {description} </PTag>
            </ContentDiv>
        </ConatinerDiv>
    )
}

export default BookDetails
