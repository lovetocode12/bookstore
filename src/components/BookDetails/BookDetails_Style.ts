import styled from "styled-components";

export const ConatinerDiv = styled.div`
    display: grid;
    grid-template-columns: 1fr 2fr;
    justify-content: center;
    align-items: flex-start;;
    width: 100%;
    height: 100%
    `

export const PhotoDiv = styled.div`
    display: flex;
    width: 80%;
    justify-content: center;
    background: #ae6060;
    `

export const Image = styled.img`
    width: 60%;
    height: 100%;
    border-radius: 2em;
    object-fit: cover;
    padding: 1em;
    `

export const ContentDiv = styled.div`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    margin: 1em 1em;
    font-size: 14px;
    height: 80%;`

export const PTag = styled.p`
    font-size: 0.8em;
    flex-grow:3;
    margin: 0;`

export const H2Tag = styled.h2`
    overflow: hidden;
    text-align: center;`

export const ButtonsContainer = styled.div`
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-grow:1;
    width: 80%;
    flex-wrap: wrap;
    margin-top: 0.4em;
`

export const StyledButton = styled.button`
    margin: 0;
    color: white;
    background: #ea8686;
    padding: 0.5em;
    font-size: 0.7rem;
    cursor: pointer;
    border-radius: 5px;
    margin-right: 1em;
    border: none;`

    export const PriceDiv = styled.div`
    
    `