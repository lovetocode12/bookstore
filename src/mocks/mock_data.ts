import { Book } from "../store/books/type";
import { Address, Order } from "../store/orders/type";

export class MockData {
    static Address: Address = {
        address1: 'address line 1',
        address2: 'address line 2',
        email: 'test@g.com',
        city: 'city',
        phone: '12345'
    }

    static Order: Order = {
        orderDate: new Date(),
        orderStatus: 'InProcess',
        book: {
            "title": "Unlocking Android",
            "id": "1933988673",
            "price": 25,
            "pageCount": 416,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/ableson.jpg",
            "description": "Unlocking Android: A Developer's Guide provides concise, hands-on instruction for the Android operating system and development tools. This book teaches important architectural concepts in a straightforward writing style and builds on this with practical and useful examples throughout.",
            "author": "W. Frank Ableson"
        }
    }
    static books: Book[] = [
        {
            "title": "Unlocking Android",
            "id": "1933988673",
            "price": 25,
            "pageCount": 416,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/ableson.jpg",
            "description": "Unlocking Android: A Developer's Guide provides concise, hands-on instruction for the Android operating system and development tools. This book teaches important architectural concepts in a straightforward writing style and builds on this with practical and useful examples throughout.",
            "author": "W. Frank Ableson"
        },
        {
            "title": "Android in Action, Second Edition",
            "id": "1935182722",
            "price": 22,
            "pageCount": 592,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/ableson2.jpg",
            "description": "Android in Action, Second Edition is a comprehensive tutorial for Android developers. Taking you far beyond \"Hello Android,\" this fast-paced book puts you in the driver's seat as you learn important architectural concepts and implementation strategies. You'll master the SDK, build WebKit apps using HTML 5, and even learn to extend or replace Android's built-in features by building useful and intriguing examples. ",
            "author": "W. Frank Ableson"
        },
        {
            "title": "Zend Framework in Action",
            "id": "1933988320",
            "price": 22,
            "pageCount": 432,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/allen.jpg",
            "description": "Zend Framework in Action is a comprehensive tutorial that shows how to use the Zend Framework to create web-based applications and web services. This book takes you on an over-the-shoulder tour of the components of the Zend Framework as you build a high quality, real-world web application.",
            "author": "Rob Allen"
        },
        {
            "title": "Flex on Java",
            "id": "1933988797",
            "pageCount": 265,
            "price": 22,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/allmon.jpg",
            "description": "   A beautifully written book that is a must have for every Java Developer.       Ashish Kulkarni, Technical Director, E-Business Software Solutions Ltd.",
            "author": "Bernerd Allmon"
        },
        {
            "title": "Griffon in Action",
            "id": "1935182234",
            "price": 22,
            "pageCount": 375,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/almiray.jpg",
            "description": "Griffon in Action is a comprehensive tutorial written for Java developers who want a more productive approach to UI development. In this book, you'll immediately dive into Griffon. After a Griffon orientation and a quick Groovy tutorial, you'll start building examples that explore Griffon's high productivity approach to Swing development. One of the troublesome parts of Swing development is the amount of Java code that is required to get a simple application off the ground.",
            "author": "Andres Almiray"
        },
        {
            "title": "OSGi in Depth",
            "id": "193518217X",
            "price": 22,
            "pageCount": 325,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/alves.jpg",
            "description": "Enterprise OSGi shows a Java developer how to develop to the OSGi Service Platform Enterprise specification, an emerging Java-based technology for developing modular enterprise applications. Enterprise OSGi addresses several shortcomings of existing enterprise platforms, such as allowing the creation of better maintainable and extensible applications, and provide a simpler, easier-to-use, light-weight solution to enterprise software development.",
            "author": "Alexandre de Castro Alves"
        },
        {
            "title": "Flexible Rails",
            "id": "1933988509",
            "price": 22,
            "pageCount": 592,
            "url": "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/alves.jpg",
            "description": "Enterprise OSGi shows a Java developer how to develop to the OSGi Service Platform Enterprise specification, an emerging Java-based technology for developing modular enterprise applications. Enterprise OSGi addresses several shortcomings of existing enterprise platforms, such as allowing the creation of better maintainable and extensible applications, and provide a simpler, easier-to-use, light-weight solution to enterprise software development.",
            "author": "Alexandre de Castro Alves"
        }
    ]

}