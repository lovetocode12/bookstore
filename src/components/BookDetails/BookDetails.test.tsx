import React from 'react';
import { shallow } from '../../../enzyme';
import BookDetails from './BookDetails';
import { MockData } from '../../mocks/mock_data';
import { H2Tag, StyledButton } from './BookDetails_Style';

jest.mock('react-redux', () => {
    const ActualReactRedux = require.requireActual('react-redux');
    return {
        ...ActualReactRedux,
        useSelector: jest.fn().mockImplementation(() => {
            return MockData.books;
        }),
        useDispatch:()=> jest.fn()
    };
});

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
    useParams: () => ({
        book_id: '1933988673'
    })
}));

const mockBuyNow = jest.fn();
jest.mock('../../common/useBuyBook', () => ({
    __esModule: true,
    default: () => ([mockBuyNow, jest.fn()])
}));


describe('BookDetails Component', () => {

    function getWrapper() {
        const wrapper: any = shallow(<BookDetails />);
        return wrapper;
    }

    it('Component should render correctly', () => {
        const wrapper = getWrapper();
        expect(wrapper).toMatchSnapshot();
    })

    it('Should display title as Flex on Java', () => {
        const wrapper = getWrapper();
        expect(wrapper.find(H2Tag).text()).toContain('Unlocking Android');
    })

    it('Should call dispatch with type BUYNOW_BOOK', () => {
        const wrapper = getWrapper();
        const buyNowButton = wrapper.find(StyledButton);
        buyNowButton.at(1).simulate('click');
        expect(mockBuyNow).toHaveBeenCalledTimes(1);
    })
});