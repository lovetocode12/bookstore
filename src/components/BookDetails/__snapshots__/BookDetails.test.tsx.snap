// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails Component Component should render correctly 1`] = `
<styled.div>
  <styled.div>
    <styled.img
      alt="Unlocking Android"
      src="https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/ableson.jpg"
    />
  </styled.div>
  <styled.div>
    <div>
      <styled.h2>
        Unlocking Android
      </styled.h2>
    </div>
    <div>
      25
    </div>
    <div>
      W. Frank Ableson
    </div>
    <div>
      416
    </div>
    <styled.div>
      <styled.button
        onClick={[Function]}
      >
        Add To Cart
      </styled.button>
      <styled.button
        onClick={[Function]}
      >
        Buy Now
      </styled.button>
    </styled.div>
    <styled.p>
       
      Unlocking Android: A Developer's Guide provides concise, hands-on instruction for the Android operating system and development tools. This book teaches important architectural concepts in a straightforward writing style and builds on this with practical and useful examples throughout.
       
    </styled.p>
  </styled.div>
</styled.div>
`;
